components/ActionSheet/src/MDCActionSheetController.m

```diff
@@ -0,0 +1,357 @@
+
+@end
+
+@implementation MDCActionSheetAction
+
++ (instancetype)actionWithTitle:(NSString *)title
+                          image:(UIImage *)image
+                          handler:(void (^__nullable)(MDCActionSheetAction *action))handler {
+    return [[MDCActionSheetAction alloc] initWithTitle:title
+                                                 image:image
+                                               handler:handler];
+}
+
+- (instancetype)initWithTitle:(NSString *)title
+                        image:(UIImage *)image
+                      handler:(void (^__nullable)(MDCActionSheetAction *action))handler {
+  self = [super init];
+  if (self) {
+    _title = [title copy];
+    _image = [image copy];
+    _completionHandler = [handler copy];
+  }
+  return self;
+}
+
+- (id)copyWithZone:(__unused NSZone *)zone {
+  MDCActionSheetAction *action = [[self class] actionWithTitle:self.title
+                                                         image:self.image
+                                                       handler:self.completionHandler];
+  return action;
+}
+
+@end
+
+@interface MDCActionSheetController () <MDCBottomSheetPresentationControllerDelegate, UITableViewDelegate>
+
+@end
+
+@implementation MDCActionSheetController {
+  MDCActionSheetHeaderView *_header;
+  UITableViewController *_tableViewController;
+  MDCActionSheetDataSource *_dataSource;
+  NSString *_actionSheetTitle;
+  MDCBottomSheetTransitionController *_transitionController;
+  BOOL mdc_adjustFontForContentSizeCategory;
+  /**
+    Used to determine if we need to do a layout within viewWillLayoutSubviews,
+    Because we are setting the preferredContentSize we can no longer get the frame to set the sheet
+    and headers sizes after a transition. 
+   */
+  BOOL needsPreferredContentSizeUpdate;
+}
+
+@synthesize mdc_adjustsFontForContentSizeCategory;
+
++ (instancetype)actionSheetControllerWithTitle:(NSString *)title message:(NSString *)message {
+  return [[MDCActionSheetController alloc] initWithTitle:title message:message];
+}
+
++ (instancetype)actionSheetControllerWithTitle:(NSString *)title {
+  return [MDCActionSheetController actionSheetControllerWithTitle:title message:nil];
+}
+
+- (instancetype)init {
+  return [MDCActionSheetController actionSheetControllerWithTitle:nil message:nil];
+}
+
+- (instancetype)initWithTitle:(NSString *)title message:(NSString *)message {
+  self = [super initWithNibName:nil bundle:nil];
+  if (self) {
+    _actionSheetTitle = [title copy];
+    _message = [message copy];
+    [self commonMDCActionSheetControllerInit];
+  }
+
+  return self;
+}
+
+- (void)commonMDCActionSheetControllerInit {
+  needsPreferredContentSizeUpdate = false;
+  _transitionController = [[MDCBottomSheetTransitionController alloc] init];
+  _transitionController.dismissOnBackgroundTap = YES;
+  /**
+   We have overriden the self implementation so we don't want others to change the values so we
+  have to call super
+   */
```

Comment:
"We must call super because we've made the setters on this class unavailable and overridden their implementations to throw assertions."

Reply:
Done